src/Misc.pm

```diff
@@ -2306,9 +2306,21 @@ sub mon_control {
 # Returns the pickupitems.txt settings for item name $name.
 # If $name has no specific settings, use 'all'.
 sub pickupitems {
-	my ($name) = @_;
-
-	return ($pickupitems{lc($name)} ne '') ? $pickupitems{lc($name)} : $pickupitems{all};
+   my $name = lc shift;
+
+   return $pickupitems{$name} if exists $pickupitems{$name} and $pickupitems{$name} ne '';
+   
+   my $result = $pickupitems{all} || {};
+   
+   foreach my $key (keys %pickupitems) {
```

Comment:
This will iterate over every item in `%pickupitems`, every time an item is considered for picking up. Compare with the hash lookup we were doing before (and are still doing for non-regex lookups), which has constant-time cost.

We should separate regexes from non-regexes in the parser, and only look at the regexes in this logic.

For comparison, here's my implementation:

```Perl
sub pickupitems {
    my $name = lc shift;
    my $pui  = $pickupitems{$name};
    $pui = ( map { $pickupitems_r{$_} } grep { $name =~ /$_/ } keys %pickupitems_r )[0] if !defined $pui;
    $pui = $pickupitems{all} if !defined $pui;
    $pui;
}
```

It uses a separate hash (`%pickupitems_r`) which only contains the regex-based rules.